Objects/iterobject.c

```diff
@@ -316,7 +316,49 @@ anextawaitable_traverse(anextawaitableobject *obj, visitproc visit, void *arg)
 static PyObject *
 anextawaitable_iternext(anextawaitableobject *obj)
 {
-    PyObject *result = PyIter_Next(obj->wrapped);
+    /* Consider the following class:
+     *
+     *     class A:
+     *         async def __anext__(self):
+     *             ...
+     *     a = A()
+     *
+     * Then `await anext(a)` should call
+     * a.__anext__().__await__().__next__()
+     *
+     * On the other hand, given
+     *
+     *     async def agen():
+     *         yield 1
+     *         yield 2
+     *     gen = agen()
+     *
+     * Then `await anext(gen)` can just call
+     * gen.__anext__().__next__()
+     */
+    assert(obj->wrapped != NULL);
+    PyObject *awaitable = _PyCoro_GetAwaitableIter(obj->wrapped);
+    if (awaitable == NULL) {
+        return NULL;
+    }
+    if (Py_TYPE(awaitable)->tp_iternext == NULL) {
+        if (Py_TYPE(awaitable)->tp_as_async == NULL ||
+            Py_TYPE(awaitable)->tp_as_async->am_await == NULL)
+        {
+            PyErr_SetString(PyExc_TypeError,
+                            "__anext__ returned a non-awaitable.");
+            return NULL;
+        }
+        unaryfunc getter = Py_TYPE(awaitable)->tp_as_async->am_await;
+        PyObject *new_awaitable = getter(awaitable);
```

Comment:
This can raise, you need to check for errors and propagate accordingly

Reply:
Also, can you add a test that checks this code path?

Reply:
I added the check, but I couldn't figure out how to make a test to check the code path, since `awaitable` is always a coroutine, so `getter` points to this in genobject.c:

```c
static PyObject *
coro_await(PyCoroObject *coro)
{
    PyCoroWrapper *cw = PyObject_GC_New(PyCoroWrapper, &_PyCoroWrapper_Type);
    if (cw == NULL) {
        return NULL;
    }
    Py_INCREF(coro);
    cw->cw_coroutine = coro;
    _PyObject_GC_TRACK(cw);
    return (PyObject *)cw;
}
```

Reply:
You need to add a test to the `_testcapi` module, but this is going to be a bit too much for just this, so is fine if we don't add such a test.

Reply:
Wait, isn't this something you can get with a custom `__await__` in a class?
 

Reply:
_PyCoro_GetAwaitableIter calls am_await and ensures it's iterable; this if block is only for coroutines.

Reply:
`test_anext_bad_await` covers `__await__` methods that raise

Reply:
👍